nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -130,6 +133,47 @@ private static void reportInvalidAssignmentInstantiationError(
             errorMessage, analysis.buildDescription(tree), state, null));
   }
 
+  private static void reportInvalidReturnTypeError(
+      Tree tree, Type methodType, Type returnType, VisitorState state, NullAway analysis) {
+    ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.RETURN_NULLABLE_GENERIC,
+            String.format(
+                "Cannot return expression of type "
+                    + returnType
+                    + " from method with return type "
+                    + methodType
+                    + " due to mismatched nullability of type parameters"));
+    state.reportMatch(
+        errorBuilder.createErrorDescription(
+            errorMessage, analysis.buildDescription(tree), state, null));
+  }
+
+  /**
+   * This method returns type of the tree considering that the parameterized typed tree annotations
+   * are not preserved if obtained directly using ASTHelpers.
```

Comment:
This javadoc doesn't quite make sense to me.

Is this approximately what you mean?

```
This method returns the type of the tree, including any type use annotations.

This is required because, in some cases, {@link 
com.google.errorprone.util.ASTHelpers#getType(Tree)} will fail to preserve the 
type use annotations. Particularly when dealing with 
{@link com.sun.source.tree.NewClassTree} (i.e. {@code new Foo<A>}).
```